syntax = "proto3";

package arduino_trader.scoring;

import "common/common.proto";
import "common/security.proto";

// Scoring Service - Handles security scoring
service ScoringService {
  // Score a single security
  rpc ScoreSecurity(ScoreSecurityRequest) returns (ScoreSecurityResponse);

  // Score multiple securities (batch)
  rpc BatchScoreSecurities(BatchScoreSecuritiesRequest) returns (BatchScoreSecuritiesResponse);

  // Score entire portfolio
  rpc ScorePortfolio(ScorePortfolioRequest) returns (ScorePortfolioResponse);

  // Get historical scores for a security
  rpc GetScoreHistory(GetScoreHistoryRequest) returns (GetScoreHistoryResponse);

  // Health check
  rpc HealthCheck(common.Empty) returns (common.HealthCheckResponse);
}

// Request to score a security
message ScoreSecurityRequest {
  string isin = 1;
  string symbol = 2;
  map<string, double> market_data = 3;
}

// Response with security score
message ScoreSecurityResponse {
  bool found = 1;
  SecurityScore score = 2;
}

// Security score
message SecurityScore {
  string isin = 1;
  string symbol = 2;
  double total_score = 3;
  map<string, double> component_scores = 4;  // momentum, value, quality, etc.
  double percentile = 5;
  string grade = 6;  // A+, A, B+, etc.
  common.Timestamp calculated_at = 7;
}

// Batch score request
message BatchScoreSecuritiesRequest {
  repeated string isins = 1;
  map<string, string> filters = 2;
}

// Batch score response
message BatchScoreSecuritiesResponse {
  repeated SecurityScore scores = 1;
  int32 total_scored = 2;
  int32 failed = 3;
}

// Portfolio score request
message ScorePortfolioRequest {
  string portfolio_hash = 1;
  repeated string isins = 2;
}

// Portfolio score response
message ScorePortfolioResponse {
  double total_score = 1;
  double weighted_score = 2;
  repeated SecurityScore security_scores = 3;
  map<string, double> portfolio_metrics = 4;
}

// Get score history request
message GetScoreHistoryRequest {
  string isin = 1;
  int32 days = 2;
}

// Get score history response
message GetScoreHistoryResponse {
  string isin = 1;
  repeated HistoricalScore scores = 2;
}

// Historical score
message HistoricalScore {
  common.Timestamp date = 1;
  double score = 2;
  string grade = 3;
}
